### ISSUE TITLE
Consider using Markdown-it plugins instead of client-libraries (highlight.js, MathJax)

### ISSUE THREAD

**Issue by MoshiKoi** [codidact/qpixel#606]
Mainly concerning https://math.codidact.com/posts/278763 and #435 but also applies to the syntax highlighting library.

Currently, we use a combination of markdown-it, highlight.js, and MathJax to parse and render posts. If I understand correctly, clients run markdown-it when writing content, converting it to HTML. Then viewers of that post run highlight.js and MathJax on it to get the result.

This has resulted in Markdown-LaTeX conflicts as linked. In order to fix that, I suggest using a markdown-it plugin such as https://www.npmjs.com/package/markdown-it-mathjax3 which integrates with the markdown parser, preventing conflicts. 

markdown-it also natively supports [using highlight-js for syntax highlighting](https://github.com/markdown-it/markdown-it#syntax-highlighting), so we could also move syntax highlighting there.

This will also alleviate some of the performance hits due to running highlight.js and MathJax client-side, because they only need to be run once on post creation, rather than whenever the user sees a page.

**Comment by ArtOfCode-** [comment]
The main hurdle to doing this is that we don't use a package system front-end, because that's how we carried over the architecture from original QPixel - we include scripts with `<script>` tags the good old way. We _could_ convert to using npm packages, but that's a pretty significant chunk of work.

**Comment by sau226** [comment]
I'm no expert at script work, but it looks like we can include it from [JSDelivr](https://www.jsdelivr.com/package/npm/markdown-it) if we reference the minified JS in their dist folder.

**Comment by MoshiKoi** [comment]
@ArtOfCode- I'm not sure what you mean by that. It's exactly the same as the footnote plugin you already use, just put it in the markdown scripts and use it.

https://github.com/codidact/qpixel/blob/f82cdf0b67299fef2abfda89e5b6ccbd2299bba8/app/assets/javascripts/posts.js#L117-L122

The main problem I see is that the scrubber will reject the rendered result

**Comment by ArtOfCode-** [comment]
@MoshiKoi From a quick skim of the readme for those packages, they want to be installed as npm packages instead of included via script tag. Haven't investigated whether they have a secondary option for script tags, but if it's just the former we can't currently do it.

**Comment by MoshiKoi** [comment]
@ArtOfCode- The one I linked doesn't have a browser version, but I've successfully gotten [another one](https://www.npmjs.com/package/@catnose99/markdown-it-texmath) to work (it's KaTeX though). 

However, what I mentioned before breaks things, I believe the scrubber is removing some of the important elements so the result doesn't look right.

**Comment by ArtOfCode-** [comment]
Yeah, I guess it would... and we can't really expand the elements it'll permit that much further without giving people tools to properly break things. It'd have to be rendered server-side, which is exactly what having the client side rendering is there to avoid.

**Comment by MoshiKoi** [comment]
@ArtOfCode- Well, you already allow the `class` attribute which can break a lot of stuff, so I think you ought to move at least some of the rendering server-side.

**Comment by ArtOfCode-** [comment]
We do have server-side rendering as a fallback, but only for cases where the client hasn't sent a rendered version. It does save fairly significant server time, so I'm reluctant to move back server-side again if we don't have to.

**Comment by MoshiKoi** [comment]
@ArtOfCode- How significant is it? I would think that the renderer would only need to run once whenever someone creates a post, which shouldn't have that much of an impact.

**Comment by ArtOfCode-** [comment]
@MoshiKoi Renders can take 100-200ms... doesn't seem huge, but it builds up. Since we _can_ get the client to do the work, we might as well.

**Comment by Istiakshovon** [comment]
I think [stackedit.io](https://stackedit.io/) is better. https://github.com/benweet/stackedit That's faster I guess.
